botocore/awsrequest.py

```diff
@@ -428,13 +425,37 @@ def reset_stream(self):
             logger.debug("Unable to rewind stream: %s", e)
             raise UnseekableStreamError(stream_object=self.body)
 
-    def prepare_body(self, data, files, json=None):
+    def prepare_url(self, url, params):
+        if params:
+            params = urlencode(list(params.items()), doseq=True)
+            self.url = '%s?%s' % (url, params)
+        else:
+            self.url = url
+
+    def prepare_headers(self, headers):
+        headers = headers or {}
+        self.headers = HeadersDict(headers.items())
+
+    def prepare_body(self, data):
         """Prepares the given HTTP body data."""
-        super(AWSPreparedRequest, self).prepare_body(data, files, json)
+        self.body = data
+
+        if self.body == b'':
+            self.body = None
+
+        if not self.body and self.method == 'GET':
+            return
+
+        if isinstance(self.body, dict):
+            params = list(self.body.items())
+            self.body = urlencode(params, doseq=True)
+
+        try:
+            length = len(self.body)
+            self.headers['Content-Length'] = str(length)
+        except Exception as e:
```

Comment:
What would raise here?  The `len()` call?

Reply:
Yep.

Reply:
Why's that?  We should scope this to something more specific than `Exception`.